hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AFailureHandling.java

```diff
@@ -72,6 +76,40 @@ public void testMultiObjectDeleteNoFile() throws Throwable {
     removeKeys(getFileSystem(), "ITestS3AFailureHandling/missingFile");
   }
 
+  /**
+   * See HADOOP-18112.
+   */
+  @Test
+  public void testMultiObjectDeleteLargeNumKeys() throws Exception {
+    S3AFileSystem fs =  getFileSystem();
+    Path path = path("largeDir");
+    mkdirs(path);
+    createFiles(fs, path, 1, 1005, 0);
+    List<String> keys = new ArrayList<>();
+    RemoteIterator<LocatedFileStatus> locatedFileStatusRemoteIterator =
+            fs.listFiles(path, false);
+    while (locatedFileStatusRemoteIterator.hasNext()) {
```

Comment:
Try using RemoteIterators here to see how well it works...do a mapping and then the toList() call

```java
toList( mappingRemoteIterator(locatedFileStatusRemoteIterator, (n) ->
 fs.pathToKey(n.getPath()));
```